src/server.rs

```diff
@@ -110,44 +111,49 @@ impl<T: Read + Write + ScmSocket> ClientConnection<T> {
+            match self.connection.try_read() {
+                Err(ConnectionError::ConnectionClosed) => {
+                    // Connection timeout.
+                    self.state = ClientConnectionState::Closed;
+                    // We don't want to propagate this to the server and we will
+                    // return no requests and wait for the connection to become
+                    // safe to drop.
+                    return Ok(vec![]);
+                }
+                Err(ConnectionError::StreamReadError(inner)) => {
+                    // Reading from the connection failed.
+                    // We should try to write an error message regardless.
+                    let mut internal_error_response =
+                        Response::new(Version::Http11, StatusCode::InternalServerError);
+                    internal_error_response.set_body(Body::new(inner.to_string()));
+                    self.connection.enqueue_response(internal_error_response);
+                    break;
+                }
+                Err(ConnectionError::ParseError(inner)) => {
+                    // An error occurred while parsing the read bytes.
+                    // Check if there are any valid parsed requests in the queue.
+                    while let Some(_discarded_request) = self.connection.pop_parsed_request() {}
+
+                    // Send an error response for the request that gave us the error.
+                    let mut error_response = Response::new(Version::Http11, StatusCode::BadRequest);
+                    error_response.set_body(Body::new(format!(
                     "{{ \"error\": \"{}\nAll previous unanswered requests will be dropped.\" }}",
                     inner
                 )));
-                self.connection.enqueue_response(error_response);
-            }
-            Err(ConnectionError::InvalidWrite) | Err(ConnectionError::StreamWriteError(_)) => {
-                // This is unreachable because `HttpConnection::try_read()` cannot return this error variant.
-                unreachable!();
-            }
-            Ok(()) => {
-                while let Some(request) = self.connection.pop_parsed_request() {
-                    // Add all valid requests to `parsed_requests`.
-                    parsed_requests.push(request);
+                    self.connection.enqueue_response(error_response);
+                    break;
+                }
+                Err(ConnectionError::InvalidWrite) | Err(ConnectionError::StreamWriteError(_)) => {
+                    // This is unreachable because `HttpConnection::try_read()` cannot return this error variant.
+                    unreachable!();
+                }
+                Ok(()) => {
+                    while let Some(request) = self.connection.pop_parsed_request() {
+                        // Add all valid requests to `parsed_requests`.
+                        parsed_requests.push(request);
+                        break 'out;
```

Comment:
Seems the `break out` should be moved out one level so we could collect all parsed reqquests.

Reply:
yes, this is been fixed in the second commit